### PR TITLE
OCPQE-20409 Phase I add Sippy API client and unit tests

### PR DESCRIPTION
**Pull request by rioliu-rh**
## Add API client with following features
- HTTP client to send Get request and support retry strategy
- URL parameter builder
  - add default values for some params
- Sippy filter builder to populate filter object list with unit (field, operator, value)
- DatetimePicker to get a datetime as endtime
- StartEndTimePicker to get from-to time range for component readiness query

## Unit test result
```console
Running tests (unittest): /Users/rio.liu/coderepo/release-tests/tests/test_sippy.py::TestSippy
Running tests: /Users/rio.liu/coderepo/release-tests/tests/test_sippy.py::TestSippy::test_health_check
/Users/rio.liu/coderepo/release-tests/tests/test_sippy.py::TestSippy::test_query_component_readiness
/Users/rio.liu/coderepo/release-tests/tests/test_sippy.py::TestSippy::test_query_jobs_by_currentruns_and_variants
/Users/rio.liu/coderepo/release-tests/tests/test_sippy.py::TestSippy::test_query_jobs_by_name
/Users/rio.liu/coderepo/release-tests/tests/test_sippy.py::TestSippy::test_query_tests
/Users/rio.liu/coderepo/release-tests/tests/test_sippy.py::TestSippy::test_query_variant
/Users/rio.liu/coderepo/release-tests/tests/test_sippy.py::TestSippy::test_start_end_time_picker
./test_sippy.py::TestSippy::test_health_check Passed

./test_sippy.py::TestSippy::test_query_component_readiness Passed

./test_sippy.py::TestSippy::test_query_jobs_by_currentruns_and_variants Passed

./test_sippy.py::TestSippy::test_query_jobs_by_name Passed

./test_sippy.py::TestSippy::test_query_tests Passed

./test_sippy.py::TestSippy::test_query_variant Passed

./test_sippy.py::TestSippy::test_start_end_time_picker Passed

Total number of tests expected to run: 7
Total number of tests run: 7
Total number of tests passed: 7
Total number of tests failed: 0
Total number of tests failed with errors: 0
Total number of tests skipped: 0

Finished running tests!
```
https://privatebin.corp.redhat.com/?43fd68042616ae2e#Crtqa2mopbWYy1bqdVWVbAZMHhauJgL43xcuUFL33ixF

### CODE REVIEW

File: prow/job/sippy.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+        if AND:
+            self._params.update(linkOperator="and")
+        else:
+            self._params.update(linkOperator="or")
+
+    def done(self):
+        if "linkOperator" not in self._params:
+            self.link()
+        self._params.update(items=self._items)
+        return json.dumps(self._params)
```
Comment (kuiwang02) [inline prow/job/sippy.py:183]: @rioliu-rh I am not sure if it encodes the filter.
for example, for filter parameter `filter={"items":[{"columnField":"failed_test_names","operatorValue":"contains","value":"[sig-network][Feature:Router][apigroup:route.openshift.io] when FIPS is enabled the HAProxy router should not work when configured with a 1024-bit RSA key [Suite:openshift/conformance/parallel]"},{"columnField":"variants","not":true,"operatorValue":"contains","value":"never-stable"},{"columnField":"variants","not":true,"operatorValue":"contains","value":"aggregated"},{"columnField":"variants","operatorValue":"contains","value":"aws"},{"columnField":"variants","operatorValue":"contains","value":"amd64"},{"columnField":"variants","operatorValue":"contains","value":"ovn"},{"columnField":"variants","operatorValue":"contains","value":"upgrade"},{"columnField":"variants","operatorValue":"contains","value":"upgrade-minor"},{"columnField":"variants","operatorValue":"contains","value":"ha"}],"linkOperator":"and"}`, it should be `filter=%7B%22items%22%3A%5B%7B%22columnField%22%3A%22failed_test_names%22%2C%22operatorValue%22%3A%22contains%22%2C%22value%22%3A%22%5Bsig-network%5D%5BFeature%3ARouter%5D%5Bapigroup%3Aroute.openshift.io%5D%20when%20FIPS%20is%20enabled%20the%20HAProxy%20router%20should%20not%20work%20when%20configured%20with%20a%201024-bit%20RSA%20key%20%5BSuite%3Aopenshift%2Fconformance%2Fparallel%5D%22%7D%2C%7B%22columnField%22%3A%22variants%22%2C%22not%22%3Atrue%2C%22operatorValue%22%3A%22contains%22%2C%22value%22%3A%22never-stable%22%7D%2C%7B%22columnField%22%3A%22variants%22%2C%22not%22%3Atrue%2C%22operatorValue%22%3A%22contains%22%2C%22value%22%3A%22aggregated%22%7D%2C%7B%22columnField%22%3A%22variants%22%2C%22operatorValue%22%3A%22contains%22%2C%22value%22%3A%22aws%22%7D%2C%7B%22columnField%22%3A%22variants%22%2C%22operatorValue%22%3A%22contains%22%2C%22value%22%3A%22amd64%22%7D%2C%7B%22columnField%22%3A%22variants%22%2C%22operatorValue%22%3A%22contains%22%2C%22value%22%3A%22ovn%22%7D%2C%7B%22columnField%22%3A%22variants%22%2C%22operatorValue%22%3A%22contains%22%2C%22value%22%3A%22upgrade%22%7D%2C%7B%22columnField%22%3A%22variants%22%2C%22operatorValue%22%3A%22contains%22%2C%22value%22%3A%22upgrade-minor%22%7D%2C%7B%22columnField%22%3A%22variants%22%2C%22operatorValue%22%3A%22contains%22%2C%22value%22%3A%22ha%22%7D%5D%2C%22linkOperator%22%3A%22and%22%7D` in request.
it should be encoded.

anyway, if you could get response with filter, it is ok.

Comment (rioliu-rh) [reply prow/job/sippy.py:183]: yes the params can be encoded by request session

**Comment by rioliu-rh** [comment]
UPDATE
add api call support for `variant` query, we can filter different variants with name and compare previous/current pass_percentage to determine whether the variant status is green

<details>
<summary> unit test result</summary>

```
$ py -m unittest test_sippy.TestSippy.test_query_variant
2024-04-01T11:10:50Z: DEBUG: Starting new HTTPS connection (1): sippy.dptools.openshift.org:443
2024-04-01T11:10:51Z: DEBUG: https://sippy.dptools.openshift.org:443 "GET /api/variants?release=4.16 HTTP/1.1" 200 None
2024-04-01T11:10:51Z: INFO: name: compact
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 0
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 50

2024-04-01T11:10:51Z: INFO: name: osd
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 0
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 0

2024-04-01T11:10:51Z: INFO: name: never-stable
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 3.9087947882736156
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 3.225806451612903

2024-04-01T11:10:51Z: INFO: name: single-node
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 27.61904761904762
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 48.24561403508772

2024-04-01T11:10:51Z: INFO: name: s390x
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 32.394366197183096
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 46.96969696969697

2024-04-01T11:10:51Z: INFO: name: metal-ipi
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 38.25503355704698
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 40.76086956521739

2024-04-01T11:10:51Z: INFO: name: serial
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 51.333333333333336
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 55.0408719346049

2024-04-01T11:10:51Z: INFO: name: libvirt
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 52.51798561151079
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 65.57377049180327

2024-04-01T11:10:51Z: INFO: name: openstack
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 59.09090909090909
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 61.36363636363637

2024-04-01T11:10:51Z: INFO: name: proxy
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 60.869565217391305
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 79.16666666666667

2024-04-01T11:10:51Z: INFO: name: metal-assisted
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 66.66666666666667
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 88.88888888888889

2024-04-01T11:10:51Z: INFO: name: vsphere-ipi
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 67.41573033707866
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 74.07407407407408

2024-04-01T11:10:51Z: INFO: name: metal-upi
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 67.6470588235294
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 76.47058823529412

2024-04-01T11:10:51Z: INFO: name: aggregated
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 68.90243902439025
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 60.869565217391305

2024-04-01T11:10:51Z: INFO: name: azure
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 69.78527607361963
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 68.10457516339869

2024-04-01T11:10:51Z: INFO: name: arm64
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 70.52631578947368
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 71.30434782608695

2024-04-01T11:10:51Z: INFO: name: vsphere-upi
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 72.41379310344827
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 55.88235294117647

2024-04-01T11:10:51Z: INFO: name: ppc64le
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 73.52941176470588
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 87.5

2024-04-01T11:10:51Z: INFO: name: sdn
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 73.9251040221914
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 74.7675962815405

2024-04-01T11:10:51Z: INFO: name: upgrade-micro
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 75.16170763260025
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 72.37991266375546

2024-04-01T11:10:51Z: INFO: name: ovn
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 75.65324316015985
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 74.62728551336146

2024-04-01T11:10:51Z: INFO: name: gcp
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 76.12612612612612
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 65.7516339869281

2024-04-01T11:10:51Z: INFO: name: amd64
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 76.61971830985915
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 75.24082270242124

2024-04-01T11:10:51Z: INFO: name: ha
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 76.63478935125356
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 75.36957558416786

2024-04-01T11:10:51Z: INFO: name: upgrade
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 76.94704049844236
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 70.4201680672269

2024-04-01T11:10:51Z: INFO: name: techpreview
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 77.03703703703704
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 69.94219653179191

2024-04-01T11:10:51Z: INFO: name: realtime
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 77.62237762237763
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 47.593582887700535

2024-04-01T11:10:51Z: INFO: name: upgrade-minor
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 78.60576923076923
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 68.35443037974683

2024-04-01T11:10:51Z: INFO: name: aws
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 80.44308632543927
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 82.88948069241012

2024-04-01T11:10:51Z: INFO: name: fips
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 83.53510895883777
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 87.22741433021807

2024-04-01T11:10:51Z: INFO: name: hypershift
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 84.30493273542601
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 87.79527559055119

2024-04-01T11:10:51Z: INFO: name: assisted
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 93.33333333333333
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 88.88888888888889

2024-04-01T11:10:51Z: INFO: name: etcd-scaling
2024-04-01T11:10:51Z: INFO: current_pass_percentage: 100
2024-04-01T11:10:51Z: INFO: previous_pass_percentage: 80

Ran 1 test in 0.856s

OK
```
</details>

File: prow/job/sippy.py (hunk 1 above)
Comment (kuiwang02) [reply prow/job/sippy.py:183]: ok, thanks

**Pull request merged**